torch/lib/c10d/ProcessGroupGloo.hpp

```diff
@@ -0,0 +1,224 @@
+  AlgorithmEntry(const AlgorithmEntry&) = delete;
+};
+
+} // namespace c10d
+
+MAKE_HASHABLE(
+    ::c10d::AlgorithmKey,
+    t.collectiveType,
+    t.type,
+    t.devices,
+    t.srcSizes,
+    t.dstSizes,
+    t.srcRank,
+    t.dstRank,
+    t.srcTensor,
+    t.dstTensor,
+    t.reduceOp);
+
+namespace c10d {
+
+// ProcessGroupGloo implements Gloo bindings for c10d.
+//
+// All functions on this class are expected to be called in the same
+// order across processes in the group. This is the only way that we
+// can guarantee to match up the same calls across processes. For
+// multi-threaded usage of process groups, you can use consider using
+// multiple process group instances.
+//
+// The Gloo algorithms that this class calls into are cached by their
+// signature (see description of AlgorithmKey above). This cache works
+// as follows: every function call instantiates an AlgorithmKey and
+// looks in the cache for existing entries. If there is one, it is
+// removed from the cache and returned to the caller. If there are
+// none, a new entry is created and returned. If an entry was created
+// before, but is still in use, the call will block and wait until the
+// entry is returned to the cache.
+//
+// In the future, we hope to extend this to allow multiple entries per
+// key, to enable parallelism for a single key. The number of entries
+// per key must always be identical for all processes. This maximum
+// number can be automatically tuned, but only if we let a single
+// process take charge, and have it broadcast the limits.
+//
+class ProcessGroupGloo : public ProcessGroup {
+ public:
+  class WorkGloo : public ProcessGroup::Work {
+   public:
+    explicit WorkGloo();
+    virtual ~WorkGloo();
+
+    // Checks if request has completed. Non-blocking operation.
+    bool isCompleted() const override;
+
+    // Waits until request completes. Blocking operation.
+    // Returns false if the work completed with an exception.
+    bool wait() override;
+
+    // Returns exception if wait() returned false.
+    const std::exception& exception() const override;
+
+   protected:
+    void finish();
+    void finishWithException(const ::gloo::Exception& ex);
+
+    std::mutex m_;
+    std::condition_variable cv_;
+    std::atomic<bool> completed_;
+
+    // Use pointer to ::gloo::Exception because it doesn't have a
+    // default constructor and constructing an empty std::unique_ptr
+    // is probably cheaper (this is highly speculative).
+    std::unique_ptr<::gloo::Exception> ex_;
+
+    friend class ProcessGroupGloo;
+  };
+
+  struct Options {
+    explicit Options();
+
+    std::vector<std::shared_ptr<::gloo::transport::Device>> devices;
+    std::chrono::milliseconds timeout;
+    int threads;
+  };
+
+  explicit ProcessGroupGloo(
+      const std::shared_ptr<Store>& store,
+      int rank,
+      int size);
+
+  virtual ~ProcessGroupGloo();
+
+  void initialize();
+
+  void initialize(Options& options);
```

Comment:
What's the reason to separate initialization from construction?

Reply:
This can be combined indeed. Not so for the destructor though. I have spent too much time on exceptions thrown in destructors masking other exceptions that the only destructor I want to deal with now is a noexcept destructor. Unfortunately destruction of this thing can throw.

Reply:
Why can the destruction throw, and how having a separate `destroy()` method helps? It's likely that this thing will be owned by something else, which will not only have to remember to call `destroy()` in its destructor, but if it will throw, then it will crash the whole program anyway, because you already are in some destructors.

Reply:
Alright, I'm changing to init on construct and destroy on destruct.

Can change again when we get bitten by it. For now it might be fine. Having control over destruction is good since then you decide when it happens, as opposed to some random object being destroying implying the process group goes away.

Reply:
I really think this is better, because we avoid having those DataChannels in invalid state. One more error that you can't make 😊 I'm pretty sure that you can defer the construction to the place where you call `initialize` and destroy them right where you would call `destruct` otherwise.